File: src/ethos/adapters/openspec/configuration.mjs
/** Observe and initialize configuration through the selected official OpenSpec package. */
import { createHash } from "node:crypto";
import { lstatSync, readFileSync, realpathSync } from "node:fs";
import { createRequire } from "node:module";
import path from "node:path";
import { pathToFileURL } from "node:url";

const [entry, expectedVersion, root, mode] = process.argv.slice(2);
const packageRoot = realpathSync(path.resolve(path.dirname(entry), ".."));
const require = createRequire(pathToFileURL(entry));
const load = (relative) => import(new URL("../dist/" + relative, pathToFileURL(entry)));
const { parse } = await import(pathToFileURL(require.resolve("yaml")));
const { readProjectConfig, resolveConfigFilePath, validateConfigRules, classifyOpenSpecDir } =
  await load("core/project-config.js");
const { resolveOpenSpecRoot } = await load("core/root-selection.js");
const { serializeConfig } = await load("core/config-prompts.js");
const { DEFAULT_OPENSPEC_SCHEMA, OPENSPEC_CONFIG_YAML, OPENSPEC_CONFIG_YML } =
  await load("core/openspec-root.js");
const { resolveSchema, getSchemaDir, listSchemas, listSchemasWithInfo } = await load(
  "core/artifact-graph/resolver.js",
);
const { loadTemplate } = await load("core/artifact-graph/instruction-loader.js");
const digest = (value) => createHash("sha256").update(value).digest("hex");
const inputs = new Map();
const warnings = [];
console.warn = (...values) => warnings.push(values.join(" "));
console.error = (...values) => warnings.push(values.join(" "));
const result = {
  path: path.join(root, OPENSPEC_CONFIG_YAML),
  content: null,
  exists: false,
  safe: true,
  default_content: serializeConfig({ schema: DEFAULT_OPENSPEC_SCHEMA }),
  verdict: "pass",
  required_gaps: [],
  warnings,
  keys: [],
  context: "",
  rules: {},
  inputs: [],
};

function nativeBinding(relative) {
  let current = root;
  const parts = relative.split("/");
  for (const [index, part] of parts.entries()) {
    current = path.join(current, part);
    let info;
    try {
      info = lstatSync(current);
    } catch (error) {
      if (error.code === "ENOENT") return;
      throw error;
    }
    if (
      info.isSymbolicLink() ||
      (index < parts.length - 1 ? !info.isDirectory() : !info.isFile())
    ) {
      result.path = path.join(root, relative);
      result.exists = true;
      result.safe = false;
      throw new Error("openspec_config_unsafe_path");
    }
  }
}

function identity(file) {
  const actual = realpathSync(file);
  for (const [prefix, base] of [
    ["package", packageRoot],
    ["repository", root],
  ]) {
    const relative = path.relative(base, actual);
    if (!relative.startsWith(".." + path.sep) && relative !== ".." && !path.isAbsolute(relative)) {
      return prefix + ":" + relative.split(path.sep).join("/");
    }
  }
  throw new Error("openspec_config_external_material");
}

function material(file) {
  const id = identity(file);
  const bytes = readFileSync(file);
  inputs.set(file, { path: id, sha256: digest(bytes) });
  return bytes;
}

try {
  const metadata = JSON.parse(readFileSync(path.join(packageRoot, "package.json"), "utf8"));
  const { program } = await import(pathToFileURL(require.resolve("@fission-ai/openspec")));
  if (
    metadata.name !== "@fission-ai/openspec" ||
    metadata.version !== expectedVersion ||
    program.version() !== expectedVersion
  ) {
    throw new Error("openspec_effective_version_mismatch");
  }
  nativeBinding(OPENSPEC_CONFIG_YAML);
  nativeBinding(OPENSPEC_CONFIG_YML);
  const selected = resolveConfigFilePath(root);
  result.exists = selected !== null;
  if (selected) {
    result.path = selected;
    result.content = new TextDecoder("utf-8", { fatal: true }).decode(material(selected));
  } else if (mode !== "initialize") {
    result.required_gaps.push("openspec_config_missing");
  }
  const raw = parse(result.content ?? result.default_content);
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    result.required_gaps.push("openspec_config_not_mapping", "openspec_config_schema_missing");
  } else {
    result.keys = Object.keys(raw);
    const config = selected ? readProjectConfig(root) : raw;
    if (!config?.schema || typeof config.schema !== "string" || !config.schema.trim()) {
      result.required_gaps.push("openspec_config_schema_missing");
    } else {
      if (selected) {
        const home = classifyOpenSpecDir(root);
        if (!home.hasPlanningShape && home.pointer.value !== undefined) {
          result.required_gaps.push("openspec_config_external_store_unsupported");
        } else {
          await resolveOpenSpecRoot({ startPath: root, allowImplicitRoot: false });
        }
      }
      const directory = getSchemaDir(config.schema, root);
      if (directory) material(path.join(directory, "schema.yaml"));
      const schema = resolveSchema(config.schema, root);
      const ids = new Set(schema.artifacts.map((artifact) => artifact.id));
      if (Object.keys(config.rules ?? {}).some((id) => !ids.has(id))) {
        for (const name of listSchemas(root)) {
          const candidate = getSchemaDir(name, root);
          if (candidate) material(path.join(candidate, "schema.yaml"));
        }
        for (const candidate of listSchemasWithInfo(root)) {
          for (const id of candidate.artifacts) ids.add(id);
        }
      }
      warnings.push(...validateConfigRules(config.rules ?? {}, ids));
      for (const artifact of schema.artifacts) {
        const file = path.join(directory, "templates", artifact.template);
        material(file);
        loadTemplate(config.schema, artifact.template, root);
      }
      result.schema = config.schema;
      result.context = config.context ?? "";
      result.rules = config.rules ?? {};
    }
  }
  for (const [file, input] of inputs) {
    if (identity(file) !== input.path || digest(readFileSync(file)) !== input.sha256) {
      result.verdict = "unknown";
      result.required_gaps.push("openspec_config_inputs_changed");
    }
  }
} catch (error) {
  const unavailable = ["EACCES", "EPERM", "EIO", "ESTALE"].includes(error.code);
  result.verdict = unavailable ? "unknown" : "block";
  result.required_gaps.push(
    unavailable
      ? "openspec_config_unavailable:" + error.code
      : error.message.startsWith("openspec_")
        ? error.message
        : "openspec_config_invalid:" + (error.diagnostic?.code ?? error.name),
  );
  result.detail = String(error);
}
if (warnings.length) result.required_gaps.push("openspec_config_native_warning");
if (result.required_gaps.length && result.verdict === "pass") result.verdict = "block";
result.inputs = [...inputs.values()].sort((a, b) =>
  a.path < b.path ? -1 : a.path > b.path ? 1 : 0,
);
result.input_digest = digest(
  JSON.stringify({
    version: expectedVersion,
    path: path.relative(root, result.path).split(path.sep).join("/"),
    exists: result.exists,
    default_content: result.default_content,
    inputs: result.inputs,
  }),
);
process.stdout.write(JSON.stringify(result));

File: src/ethos/adapters/openspec/configuration.py
"""Observe native OpenSpec configuration and schema inputs without host-global state."""

from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path
from tempfile import TemporaryDirectory

from ethos.adapters.openspec.cli import OFFICIAL_VERSION
from ethos.adapters.openspec.cli import openspec_base_command
from ethos.adapters.process import ProcessExecutionError
from ethos.adapters.process import run_command
from ethos.normalization.coercion import string_sequence
from ethos.repository.openspec.audit import configuration_policy_gaps


def official_config_report(root: Path, *, initialize: bool = False) -> dict[str, object]:
    """Resolve the selected native config, schema and templates before claiming readiness."""
    target = root.resolve()
    command = openspec_base_command(execution_probe=False)
    fallback: dict[str, object] = {
        "verdict": "unknown",
        "path": str(target / "openspec/config.yaml"),
        "content": None,
        "default_content": "",
        "exists": False,
        "safe": False,
        "required_gaps": ["openspec_configuration_observation_unavailable"],
        "inputs": [],
        "input_digest": "",
    }
    if command is None:
        return fallback | {"required_gaps": ["openspec_official_cli_missing"]}
    try:
        with TemporaryDirectory(prefix="ethos-openspec-configuration-") as temporary:
            environment = {
                key: os.environ[key]
                for key in ("PATH", "SystemRoot", "WINDIR")
                if key in os.environ
            }
            environment.update(
                HOME=temporary,
                USERPROFILE=temporary,
                XDG_DATA_HOME=temporary,
                XDG_CONFIG_HOME=temporary,
                OPENSPEC_TELEMETRY="0",
                OPENSPEC_NO_UPDATE_CHECK="1",
            )
            completed = run_command(
                target,
                (
                    command[0],
                    str(Path(__file__).with_suffix(".mjs")),
                    command[1],
                    OFFICIAL_VERSION,
                    str(target),
                    "initialize" if initialize else "inspect",
                ),
                env=environment,
                inherit_environment=False,
                timeout=30,
            )
        payload = json.loads(completed.stdout)
        if (
            completed.returncode
            or completed.stderr
            or not isinstance(payload, dict)
            or payload.get("verdict") not in {"pass", "block", "unknown"}
            or not isinstance(payload.get("path"), str)
            or not isinstance(payload.get("required_gaps"), list)
            or not all(isinstance(gap, str) for gap in payload["required_gaps"])
            or not isinstance(payload.get("input_digest"), str)
            or not isinstance(payload.get("default_content"), str)
            or not isinstance(payload.get("content"), (str, type(None)))
            or not isinstance(payload.get("exists"), bool)
            or not isinstance(payload.get("safe"), bool)
            or not isinstance(payload.get("inputs"), list)
            or not Path(payload["path"]).is_relative_to(target)
        ):
            return fallback | {"detail": completed.stderr or completed.stdout}
    except (OSError, ProcessExecutionError, subprocess.TimeoutExpired, ValueError) as error:
        return fallback | {"detail": f"{type(error).__name__}: {error}"}
    payload["required_gaps"].extend(configuration_policy_gaps(string_sequence(payload.get("keys"))))
    if payload["required_gaps"] and payload["verdict"] == "pass":
        payload["verdict"] = "block"
    return payload
